Move R tests to GitHub Actions

lorenzwalthert opened the pull request:
## What changes are proposed in this pull request?

Move the R tests from Travis to GitHub actions. This was previously attempted unsuccessfully because a missing `libcurl` system requirement was not installed in https://github.com/mlflow/mlflow/pull/2724, so I am basically completing some of @aarondav's work here. This closes #3047. Also, it is supposed to unblock #2367 because of newer gcc versions.

## How is this patch tested?

Github Actions 😊. I mainly copied the scripts, so the same should be tested as we tested on Travis. I managed to cache R package installation as described in https://github.com/r-lib/actions/tree/master/examples and code coverage also seems to work.

Some things up for debate as far as the CI specification goes are:

- R versions (currently only latest).
- platform: currently only ubuntu.
- platform version: currently only 18.04. 16.04 did also work (see 0176c59).

I am unsure how restricted you are in terms of build time. Please let me know if I should add more builds to the matrix.

TODO before merge

- Re-add all non-R tests. I disabled them for testing the R build. 

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.


### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [x] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

aarondav committed:
Run all small tests using GitHub Actions

aarondav committed:
Fix R & Java

aarondav committed:
Fix yaml

aarondav committed:
Fix Java

aarondav committed:
Devtools

aarondav committed:
Try different process to install R

lorenzwalthert committed:
merge from master

lorenzwalthert committed:
use latest R, not (broken) v1

lorenzwalthert committed:
don't use latest ubuntu

lorenzwalthert committed:
downgrade R

lorenzwalthert committed:
try lower version to force binary

lorenzwalthert committed:
force binary package

lorenzwalthert committed:
try to install system deps

lorenzwalthert committed:
install sys deps

lorenzwalthert committed:
only run r tests for now

lorenzwalthert committed:
try other libcurl

lorenzwalthert committed:
set root dir of R package

lorenzwalthert committed:
more verbose about conda envs

lorenzwalthert committed:
try to cache R packages

lorenzwalthert committed:
better caching

lorenzwalthert committed:
don't use pkg manager

lorenzwalthert committed:
cut script off to find error

lorenzwalthert committed:
git root is not R package root

lorenzwalthert committed:
redundant quote

lorenzwalthert committed:
again test package

lorenzwalthert committed:
Add code coverage

lorenzwalthert committed:
cd again

lorenzwalthert committed:
select right directory

lorenzwalthert committed:
also set env explicitly for now, maybe drop the heuristic to derive it later, becuase it depends on from where the script .travis.R is called called

lorenzwalthert committed:
remove R from travis

lorenzwalthert committed:
drop ambiguity about os

lorenzwalthert committed:
move back to 18.04, as 16.04 fails

smurching commented:
@lorenzwalthert this is fantastic! Thanks for the contribution, and let us know when it's ready for review :)

smurching commented:
BTW, I think testing the latest R version on Ubuntu 18.04 as you've done should suffice for now - I don't think we use any platform-specific R APIs (or unstable R APIs) that would introduce significant risk if we don't test against other OSes or R versions

lorenzwalthert committed:
add non-R tests again

lorenzwalthert committed:
don't pin R version

lorenzwalthert committed:
don't cache on 'release', but on actual R version to invalidate the cache when 'release' is bumped

lorenzwalthert committed:
- don't create matrix as we only need one entry
- rely on ubuntu-latest instead of 18.04 explicitly
- remove random comment
- rename action from r-tests to r to match other actions

lorenzwalthert commented:
Ok, thanks @smurching. I think this is ready for merge. For some reason, the lintr action for python now does not pass anymore, but it's unrelated to what I did, so I hope we can merge this soon. I might do some more work on the R CI/CD to make it a bit cleaner after #2367.

lorenzwalthert committed:
Merge branch 'master' into gh-actions

lorenzwalthert committed:
Merge branch 'master' into gh-actions

lorenzwalthert commented:
@smurching all lights green after merging master into this branch. I think this is ready to merge from my side.

smurching commented on .github/workflows/master.yml:
```diff
@@ -27,7 +27,50 @@ jobs:
 
+    - name: Cache R packages
+      if: runner.os != 'Windows'
+      uses: actions/cache@v1
```
Looks like the latest version of the `cache` action is v2: https://github.com/actions/cache#whats-new

Out of curiosity, any reason we're using v1 here? If not, maybe we could just upgrade to v2 preemptively

lorenzwalthert replied:
right. TBH, I just copied from the template from r-lib/actions. Let's try `v2`.

smurching commented on .github/workflows/master.yml:
```diff
@@ -27,7 +27,50 @@ jobs:
+    steps:
+    - uses: actions/checkout@master
+    - uses: r-lib/actions/setup-r@v1
+    - name: Query dependencies
```
```suggestion
    # This step dumps the current set of R dependencies and R version into files to be used
    # as a cache key when caching/restoring R dependencies.
    - name: Query dependencies
```

Would be good to add a clarifying comment explaining the purpose of this step

lorenzwalthert replied:
Sure. Also, we might want to use the `r-cache` action (https://github.com/r-lib/actions/issues/114) once it's implemented.

smurching replied:
Yep makes sense!

smurching commented on .github/workflows/master.yml:
```diff
@@ -27,7 +27,50 @@ jobs:
+      uses: actions/cache@v1
+      with:
+        path: ${{ env.R_LIBS_USER }}
+        key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
```
```suggestion
        # We cache R dependencies based on a tuple of the current OS, the R version, and the list of
        # R dependencies 
        key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
```

smurching reviewed (commented):
Some small comments, otherwise largely LGTM - thanks @lorenzwalthert!

lorenzwalthert committed:
more verbose explanation

Co-authored-by: Siddharth Murching <smurching@gmail.com>

lorenzwalthert committed:
more verbose explanation for caching

Co-authored-by: Siddharth Murching <smurching@gmail.com>

lorenzwalthert committed:
use v2 for caching

lorenzwalthert commented:
Thanks @smurching, I adapted to your suggestions.

smurching reviewed (approved):
LGTM

smurching commented:
Thanks again @lorenzwalthert!

javierluraschi commented:
Thank you, @lorenzwalthert  and @smurching!
